src/test/regress/upgrade/config.py

```diff
@@ -1,68 +1,277 @@
+            cur_worker_name = "worker{}".format(i)
+            self.node_name_to_ports[cur_worker_name] = cur_port
+            self.worker_ports.append(cur_port)
+
+    def _get_and_update_next_port(self):
+        if hasattr(self, "fixed_port"):
+            next_port = self.fixed_port
+            self.fixed_port += 1
+            return next_port
+        return find_free_port()
+
+
+class CitusMXBaseClusterConfig(CitusBaseClusterConfig):
     def __init__(self, arguments):
-        self.old_bindir = arguments['--old-bindir']
-        self.new_bindir = arguments['--new-bindir']
-        self.pg_srcdir = arguments['--pgxsdir']
-        self.temp_dir = './tmp_upgrade'
-        self.old_datadir = self.temp_dir + '/oldData'
-        self.new_datadir = self.temp_dir + '/newData'
-        self.settings = {
-            'shared_preload_libraries': 'citus',
-            'citus.node_conninfo': 'sslmode=prefer'
+        super().__init__(arguments)
+        self.is_mx = True
+        self.add_coordinator_to_metadata = True
+
+
+class CitusUpgradeConfig(CitusBaseClusterConfig):
+    def __init__(self, arguments):
+        super().__init__(arguments)
+        self.pre_tar_path = arguments["--citus-pre-tar"]
+        self.post_tar_path = arguments["--citus-post-tar"]
+        self.temp_dir = "./tmp_citus_upgrade"
+        self.new_settings = {"citus.enable_version_checks": "false"}
+        self.user = SUPER_USER_NAME
+        self.mixed_mode = arguments["--mixed"]
+        self.fixed_port = 57635
+
+
+class CitusDefaultClusterConfig(CitusBaseClusterConfig):
+    pass
+
+
+class CitusSingleNodeClusterConfig(CitusDefaultClusterConfig):
+    def __init__(self, arguments):
+        super().__init__(arguments)
+        self.worker_amount = 0
+
+
+class CitusSingleWorkerClusterConfig(CitusMXBaseClusterConfig):
+    def __init__(self, arguments):
+        super().__init__(arguments)
+        self.worker_amount = 1
+
+
+class CitusSuperUserDefaultClusterConfig(CitusMXBaseClusterConfig):
+    def __init__(self, arguments):
+        super().__init__(arguments)
+        self.user = SUPER_USER_NAME
+
+
+class CitusFiveWorkersManyShardsClusterConfig(CitusMXBaseClusterConfig):
+    def __init__(self, arguments):
+        super().__init__(arguments)
+        self.new_settings = {"citus.shard_count": 494}
+        self.worker_amount = 5
+
+    def setup_steps(self):
+        common.coordinator_should_haveshards(self.bindir, self.coordinator_port())
+
+
+class CitusSmallSharedPoolSizeConfig(CitusMXBaseClusterConfig):
+    def __init__(self, arguments):
+        super().__init__(arguments)
+        self.new_settings = {
+            "citus.local_shared_pool_size": 2,
+            "citus.max_shared_pool_size": 2,
+        }
+
+
+class CitusSmallExecutorPoolSizeConfig(CitusMXBaseClusterConfig):
+    def __init__(self, arguments):
+        super().__init__(arguments)
+        self.new_settings = {
+            "citus.max_adaptive_executor_pool_size": 2,
```

Comment:
With MX and repartitioning, max_adaptive_executor_pool_size = 2 is can be too small as repartitioning doesn't obey `max_adaptive_executor_pool_size` . I think making this value something like ~5 would be good enough. 

We already have tests making sure that we act properly when the pool size full. This is to make sure that even if the pool size is too low, some connections wait, we can still finish.